APKBUILD.5

```diff
@@ -0,0 +1,488 @@
+You must implement this function yourself.  If no compilation is required, you
+may simply write
+.Li return 0
+to skip the build phase.
+
+.It Cm check
+Runs the package's test suite.  This function must be implemented unless
+.Li Dq !check
+was specified in
+.Cm options .
+
+.It Cm package
+Installs the package into
+.Ev pkgdir .
+Note that
+.Ev pkgdir
+is not created for you; if this package installs no files (for example, a
+metapackage), you must use
+.Li mkdir -p "$pkgdir"
+to skip the package phase.
+
+.El
+
+
+.Ss Install Scripts
+An install script is run when an action is taken on a package by
+.Xr apk 8 .
+An install script must be written in shell and must have a
+.Li Dq #!/bin/sh
+interpreter declaration as the first line.  The
+.Cm install
+variable must contain the install scripts needed by the package.  The different
+actions available are as follows:
+
+.Bl -tag -width Ds
+.It Ic $pkgname.pre-install
+Executed before the package is installed.  If this script exits with an error
+(non-zero exit code),
+.Xr apk 8
+will halt the installation and the package will not be installed.  This install
+script is typically used to create any users or groups needed as described in
+.Cm pkggroups
+and
+.Cm pkgusers .
+
+.It Ic $pkgname.post-install
+Executed after the package is installed.  If this script exits with an error
+(non-zero exit code),
+.Xr apk 8
+will mark the package as broken.  The
+.Li apk fix
+command will attempt to re-run the post-install script if this occurs.
+
+.It Ic $pkgname.pre-upgrade
+Executed before the package is upgraded.  If this script exits with an error
+(non-zero exit code),
+.Xr apk 8
+will mark the package as broken.
+
+.It Ic $pkgname.post-upgrade
+Executed after the package is upgraded.  If this script exits with an error
+(non-zero exit code),
+.Xr apk 8
+will mark the package as broken.  The
+.Li apk fix
+command will attempt to re-run the post-upgrade script if this occurs.
+
+.It Ic $pkgname.pre-deinstall
+Executed before the package is removed from the system.  If this script exits
+with an error (non-zero exit code),
+.Xr apk 8
+will not remove the package from the system.
+
+.It Ic $pkgname.post-deinstall
+Executed after the package is removed from the system.  Exiting with an error
+will have no effect.
+
+.El
+
+
+.Sh IMPLEMENTATION NOTES
+Currently,
+.Nm
+files are sourced as normal shell scripts.  This may change at a later date.
+The
+.Xr abuild 1
+utility as distributed by Alpine uses the ash shell, a part of
+.Xr busybox 1
+that is currently undocumented.  It is mostly compliant with
+.St -p1003.2 ,
+but this cannot be fully relied upon.  The
+.Xr abuild 1
+utility as distributed by Adélie uses whatever the user has set as /bin/sh,
+which is typically
+.Xr bash 1
+or
+.Xr dash 1
+on Adélie systems.
+This is something to keep in mind as you write more complex packages.
+
+
+.Sh COMPATIBILITY
+It is highly recommended that package functions comply with
+.St -p1003.2
+for maximum compatibility.
```

Comment:
This is maybe inaccurate. We require using some specific non-POSIX features/extensions, such as `local`, and prefer using e.g. `${var//x/y}` instead of subshell outside of function (to avoid slowdown when evaluating abuilds just to read variables) or insane workarounds.

Reply:
[`abuild` does not work on dash](https://lists.adelielinux.org/hyperkitty/list/adelie-devel@lists.adelielinux.org/message/OXFARB4W4UGR3SZRLRCOKCTFYQSP7H3Y/).

I can remove this section.

Reply:
Aha, we should probably fix that – I think that `abuild` itself should be compatible with common POSIX-sh shells, including dash.
However, I’d disagree with requiring this compatibility (i.e. not using `${var//x/y}`) in APKBUILDs, because it’d considerably harm readability/maintainability of many APKBUILDs.

BTW `abuild` is already quite huge monolithic shell script doing too many things. It’s also quite slow. There are some parts, such as scanning shared objects for dependencies/provides, that IMHO should be extracted from it and rewritten into Lua or C tool. WDYT?

Reply:
As i wrote on adelie-devel when discussing `/bin/dash` as `/bin/sh`, I believe that we should have abuild working on any Almquist shell.

I believe NetBSD ash does support `${var//x/y}` substitutions, so the solution is probably to fix dash.

Reply:
> …the solution is probably to fix dash.

That’d be great!